// react-image-magnify/dist/es/ReactImageMagnify.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import detectIt from 'detect-it';
import objectAssign from 'object-assign';
import PropTypes from 'prop-types';
import React from 'react';
import ReactCursorPosition from 'react-cursor-position';
import requiredIf from 'react-required-if';

import DisplayUntilActive from './hint/DisplayUntilActive';
import EnlargedImage from './EnlargedImage';
import { getLensCursorOffset } from './lib/lens';
import Hint from './hint/DefaultHint';
import ShadedLens from './shaded-lens';
import ImageShape from './prop-types/ImageShape';
import { noop } from './utils';
import { INPUT_TYPE, ENLARGED_IMAGE_POSITION } from './constants';

var ReactImageMagnify = function (_React$Component) {
    _inherits(ReactImageMagnify, _React$Component);

    function ReactImageMagnify(props) {
        _classCallCheck(this, ReactImageMagnify);

        var _this = _possibleConstructorReturn(this, (ReactImageMagnify.__proto__ || Object.getPrototypeOf(ReactImageMagnify)).call(this, props));

        var primaryInput = detectIt.primaryInput;
        var MOUSE = INPUT_TYPE.mouse,
            TOUCH = INPUT_TYPE.touch;


        _this.state = {
            smallImageWidth: 0,
            smallImageHeight: 0,
            detectedInputType: {
                isMouseDeteced: primaryInput === MOUSE,
                isTouchDetected: primaryInput === TOUCH
            },
            isActive: false
        };

        _this.onSmallImageLoad = _this.onSmallImageLoad.bind(_this);
        _this.setSmallImageDimensionState = _this.setSmallImageDimensionState.bind(_this);
        _this.onDetectedInputTypeChanged = _this.onDetectedInputTypeChanged.bind(_this);
        _this.onActivationChanged = _this.onActivationChanged.bind(_this);
        return _this;
    }

    _createClass(ReactImageMagnify, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            var isFluidWidth = this.props.smallImage.isFluidWidth;


            if (!isFluidWidth) {
                return;
            }

            this.setSmallImageDimensionState();
            window.addEventListener('resize', this.setSmallImageDimensionState);
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            window.removeEventListener('resize', this.setSmallImageDimensionState);
        }
    }, {
        key: 'onSmallImageLoad',
        value: function onSmallImageLoad(e) {
            var _props$smallImage$onL = this.props.smallImage.onLoad,
                onLoad = _props$smallImage$onL === undefined ? noop : _props$smallImage$onL;


            onLoad(e);

            if (!this.props.smallImage.isFluidWidth) {
                return;
            }

            this.setSmallImageDimensionState();
        }
    }, {
        key: 'setSmallImageDimensionState',
        value: function setSmallImageDimensionState() {
            var _smallImageEl = this.smallImageEl,
                smallImageWidth = _smallImageEl.offsetWidth,
                smallImageHeight = _smallImageEl.offsetHeight;


            this.setState({
                smallImageWidth: smallImageWidth,
                smallImageHeight: smallImageHeight
            });
        }
    }, {
        key: 'onDetectedInputTypeChanged',
        value: function onDetectedInputTypeChanged(detectedInputType) {
            this.setState({
                detectedInputType: detectedInputType
            });
        }
    }, {
        key: 'onActivationChanged',
        value: function onActivationChanged(_ref) {
            var isActive = _ref.isActive;

            this.setState({
                isActive: isActive
            });
        }
    }, {
        key: 'getEnlargedImagePlacement',
        value: function getEnlargedImagePlacement() {
            var userDefinedEnlargedImagePosition = this.props.enlargedImagePosition;
            var isTouchDetected = this.state.detectedInputType.isTouchDetected;

            var computedEnlargedImagePosition = isTouchDetected ? ENLARGED_IMAGE_POSITION.over : ENLARGED_IMAGE_POSITION.beside;

            return userDefinedEnlargedImagePosition || computedEnlargedImagePosition;
        }
    }, {
        key: 'render',
        value: function render() {
            var _this2 = this;

            var _props = this.props,
                className = _props.className,
                enlargedImageContainerClassName = _props.enlargedImageContainerClassName,
                enlargedImageContainerStyle = _props.enlargedImageContainerStyle,
                enlargedImageClassName = _props.enlargedImageClassName,
                enlargedImageStyle = _props.enlargedImageStyle,
                fadeDurationInMs = _props.fadeDurationInMs,
                HintComponent = _props.hintComponent,
                shouldHideHintAfterFirstActivation = _props.shouldHideHintAfterFirstActivation,
                isHintEnabled = _props.isHintEnabled,
                hintTextMouse = _props.hintTextMouse,
                hintTextTouch = _props.hintTextTouch,
                hoverDelayInMs = _props.hoverDelayInMs,
                hoverOffDelayInMs = _props.hoverOffDelayInMs,
                isActivatedOnTouch = _props.isActivatedOnTouch,
                imageClassName = _props.imageClassName,
                imageStyle = _props.imageStyle,
                largeImage = _props.largeImage,
                lensStyle = _props.lensStyle,
                pressDuration = _props.pressDuration,
                pressMoveThreshold = _props.pressMoveThreshold,
                _props$smallImage = _props.smallImage,
                isSmallImageFluidWidth = _props$smallImage.isFluidWidth,
                _props$smallImage$onE = _props$smallImage.onError,
                onError = _props$smallImage$onE === undefined ? noop : _props$smallImage$onE,
                style = _props.style;
            var _state = this.state,
                smallImageWidth = _state.smallImageWidth,
                smallImageHeight = _state.smallImageHeight,
                isTouchDetected = _state.detectedInputType.isTouchDetected;


            var fluidWidthSmallImage = objectAssign({}, this.props.smallImage, {
                width: smallImageWidth,
                height: smallImageHeight
            });

            var fixedWidthSmallImage = this.props.smallImage;

            var smallImage = isSmallImageFluidWidth ? fluidWidthSmallImage : fixedWidthSmallImage;

            var fluidWidthContainerStyle = {
                width: 'auto',
                height: 'auto',
                fontSize: '0px',
                position: 'relative'
            };
            var fixedWidthContainerStyle = {
                width: smallImage.width + 'px',
                height: smallImage.height + 'px',
                position: 'relative'
            };
            var priorityContainerStyle = isSmallImageFluidWidth ? fluidWidthContainerStyle : fixedWidthContainerStyle;
            var compositContainerStyle = objectAssign({
                cursor: 'crosshair'
            }, style, priorityContainerStyle);

            var fluidWidthSmallImageStyle = {
                width: '100%',
                height: 'auto',
                display: 'block',
                pointerEvents: 'none'
            };
            var fixedWidthSmallImageStyle = {
                width: smallImage.width + 'px',
                height: smallImage.height + 'px',
                pointerEvents: 'none'
            };
            var prioritySmallImageStyle = isSmallImageFluidWidth ? fluidWidthSmallImageStyle : fixedWidthSmallImageStyle;
            var compositSmallImageStyle = objectAssign({}, imageStyle, prioritySmallImageStyle);

            var enlargedImagePlacement = this.getEnlargedImagePlacement();

            var shouldShowLens = enlargedImagePlacement !== ENLARGED_IMAGE_POSITION.over && !isTouchDetected;

            var cursorOffset = getLensCursorOffset(smallImage, largeImage);

            return React.createElement(
                ReactCursorPosition,
                {
                    className: className,
                    hoverDelayInMs: hoverDelayInMs,
                    hoverOffDelayInMs: hoverOffDelayInMs,
                    isActivatedOnTouch: isActivatedOnTouch,
                    onActivationChanged: this.onActivationChanged,
                    onDetectedInputTypeChanged: this.onDetectedInputTypeChanged,
                    pressDuration: pressDuration,
                    pressMoveThreshold: pressMoveThreshold,
                    shouldStopTouchMovePropagation: true,
                    style: compositContainerStyle
                },
                React.createElement('img', {
                    src: smallImage.src,
                    srcSet: smallImage.srcSet,
                    sizes: smallImage.sizes,
                    alt: smallImage.alt,
                    className: imageClassName,
                    style: compositSmallImageStyle,
                    ref: function ref(el) {
                        return _this2.smallImageEl = el;
                    },
                    onLoad: this.onSmallImageLoad,
                    onError: onError
                }),
                isHintEnabled && React.createElement(
                    DisplayUntilActive,
                    {
                        shouldHideAfterFirstActivation: shouldHideHintAfterFirstActivation
                    },
                    React.createElement(HintComponent, {
                        isTouchDetected: isTouchDetected,
                        hintTextMouse: hintTextMouse,
                        hintTextTouch: hintTextTouch
                    })
                ),
                shouldShowLens && React.createElement(ShadedLens, {
                    cursorOffset: cursorOffset,
                    fadeDurationInMs: fadeDurationInMs,
                    smallImage: smallImage,
                    style: lensStyle
                }),
                React.createElement(EnlargedImage, {
                    containerClassName: enlargedImageContainerClassName,
                    containerStyle: enlargedImageContainerStyle,
                    cursorOffset: cursorOffset,
                    fadeDurationInMs: fadeDurationInMs,
                    imageClassName: enlargedImageClassName,
                    imageStyle: enlargedImageStyle,
                    imagePosition: enlargedImagePlacement,
                    largeImage: largeImage,
                    smallImage: smallImage
                })
            );
        }
    }]);

    return ReactImageMagnify;
}(React.Component);

ReactImageMagnify.propTypes = {
    className: PropTypes.string,
    enlargedImageContainerClassName: PropTypes.string,
    enlargedImageContainerStyle: PropTypes.object,
    enlargedImageClassName: PropTypes.string,
    enlargedImageStyle: PropTypes.object,
    fadeDurationInMs: PropTypes.number,
    hintComponent: PropTypes.func,
    shouldHideHintAfterFirstActivation: PropTypes.bool,
    isHintEnabled: PropTypes.bool,
    hintTextMouse: PropTypes.string,
    hintTextTouch: PropTypes.string,
    hoverDelayInMs: PropTypes.number,
    hoverOffDelayInMs: PropTypes.number,
    isActivatedOnTouch: PropTypes.bool,
    imageClassName: PropTypes.string,
    imageStyle: PropTypes.object,
    largeImage: ImageShape,
    lensStyle: PropTypes.object,
    pressDuration: PropTypes.number,
    pressMoveThreshold: PropTypes.number,
    smallImage: PropTypes.shape({
        alt: PropTypes.string,
        isFluidWidth: PropTypes.bool,
        src: PropTypes.string.isRequired,
        srcSet: PropTypes.string,
        sizes: PropTypes.string,
        width: requiredIf(PropTypes.number, function (props) {
            return !props.isFluidWidth;
        }),
        height: requiredIf(PropTypes.number, function (props) {
            return !props.isFluidWidth;
        }),
        onLoad: PropTypes.func,
        onError: PropTypes.func
    }),
    style: PropTypes.object,
    enlargedImagePosition: PropTypes.oneOf([ENLARGED_IMAGE_POSITION.beside, ENLARGED_IMAGE_POSITION.over])
};
ReactImageMagnify.defaultProps = {
    fadeDurationInMs: 300,
    hintComponent: Hint,
    shouldHideHintAfterFirstActivation: true,
    isHintEnabled: false,
    hintTextMouse: 'Hover to Zoom',
    hintTextTouch: 'Long-Touch to Zoom',
    hoverDelayInMs: 250,
    hoverOffDelayInMs: 150
};


export default ReactImageMagnify;
