// react-image-magnify/dist/EnlargedImage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _imageCoordinates = require('./lib/imageCoordinates');

var _ImageShape = require('./prop-types/ImageShape');

var _ImageShape2 = _interopRequireDefault(_ImageShape);

var _utils = require('./utils');

var _Point = require('./prop-types/Point');

var _Point2 = _interopRequireDefault(_Point);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _class = function (_React$Component) {
    _inherits(_class, _React$Component);

    function _class(props) {
        _classCallCheck(this, _class);

        var _this = _possibleConstructorReturn(this, (_class.__proto__ || Object.getPrototypeOf(_class)).call(this, props));

        _this.state = {
            isTransitionEntering: false,
            isTransitionActive: false,
            isTransitionLeaving: false,
            isTransitionDone: false
        };

        _this.timers = [];
        return _this;
    }

    _createClass(_class, [{
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            this.scheduleCssTransition(nextProps);
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            this.timers.forEach(function (timerId) {
                clearTimeout(timerId);
            });
        }
    }, {
        key: 'scheduleCssTransition',
        value: function scheduleCssTransition(nextProps) {
            var _this2 = this;

            var _props = this.props,
                fadeDurationInMs = _props.fadeDurationInMs,
                isActive = _props.isActive,
                isPositionOutside = _props.isPositionOutside;

            var willIsActiveChange = isActive !== nextProps.isActive;
            var willIsPositionOutsideChange = isPositionOutside !== nextProps.isPositionOutside;

            if (!willIsActiveChange && !willIsPositionOutsideChange) {
                return;
            }

            if (nextProps.isActive && !nextProps.isPositionOutside) {
                this.setState({
                    isTrainsitionDone: false,
                    isTransitionEntering: true
                });

                this.timers.push(setTimeout(function () {
                    _this2.setState({
                        isTransitionEntering: false,
                        isTransitionActive: true
                    });
                }, 0));
            } else {
                this.setState({
                    isTransitionLeaving: true,
                    isTransitionActive: false
                });

                this.timers.push(setTimeout(function () {
                    _this2.setState({
                        isTransitionDone: true,
                        isTransitionLeaving: false
                    });
                }, fadeDurationInMs));
            }
        }
    }, {
        key: 'getDefaultContainerStyle',
        value: function getDefaultContainerStyle() {
            var imagePosition = this.props.imagePosition;

            var baseContainerStyle = {
                position: 'absolute',
                top: '0px',
                overflow: 'hidden'
            };
            var OVER = _constants.ENLARGED_IMAGE_POSITION.over;

            var isInPlaceMode = imagePosition === OVER;

            if (isInPlaceMode) {
                return (0, _objectAssign2.default)(baseContainerStyle, {
                    left: '0px'
                });
            }

            return (0, _objectAssign2.default)(baseContainerStyle, {
                left: '100%',
                marginLeft: '10px',
                border: '1px solid #d6d6d6'
            });
        }
    }, {
        key: 'getImageCoordinates',
        value: function getImageCoordinates() {
            var _props2 = this.props,
                imagePosition = _props2.imagePosition,
                cursorOffset = _props2.cursorOffset,
                largeImage = _props2.largeImage,
                smallImage = _props2.smallImage,
                position = _props2.position;
            var OVER = _constants.ENLARGED_IMAGE_POSITION.over;

            var isInPlaceMode = imagePosition === OVER;

            if (isInPlaceMode) {
                return (0, _imageCoordinates.getInPlaceEnlargedImageCoordinates)(smallImage, largeImage, position);
            }

            return (0, _imageCoordinates.getLensModeEnlargedImageCoordinates)(smallImage, largeImage, position, cursorOffset);
        }
    }, {
        key: 'render',
        value: function render() {
            var _props3 = this.props,
                containerClassName = _props3.containerClassName,
                containerStyle = _props3.containerStyle,
                fadeDurationInMs = _props3.fadeDurationInMs,
                imageClassName = _props3.imageClassName,
                imageStyle = _props3.imageStyle,
                isLazyLoaded = _props3.isLazyLoaded,
                largeImage = _props3.largeImage,
                _props3$largeImage = _props3.largeImage,
                _props3$largeImage$on = _props3$largeImage.onLoad,
                onLoad = _props3$largeImage$on === undefined ? _utils.noop : _props3$largeImage$on,
                _props3$largeImage$on2 = _props3$largeImage.onError,
                onError = _props3$largeImage$on2 === undefined ? _utils.noop : _props3$largeImage$on2,
                smallImage = _props3.smallImage;
            var _state = this.state,
                isTransitionEntering = _state.isTransitionEntering,
                isTransitionActive = _state.isTransitionActive,
                isTransitionLeaving = _state.isTransitionLeaving;

            var isVisible = !!(isTransitionEntering || isTransitionActive || isTransitionLeaving);

            var defaultContainerStyle = this.getDefaultContainerStyle();
            var computedContainerStyle = {
                width: smallImage.width,
                height: smallImage.height,
                opacity: this.state.isTransitionActive ? 1 : 0,
                transition: 'opacity ' + fadeDurationInMs + 'ms ease-in',
                pointerEvents: 'none'
            };
            var compositeContainerStyle = (0, _objectAssign2.default)({}, defaultContainerStyle, containerStyle, computedContainerStyle);

            var imageCoordinates = this.getImageCoordinates();
            var translate = 'translate(' + imageCoordinates.x + 'px, ' + imageCoordinates.y + 'px)';
            var computedImageStyle = {
                width: largeImage.width,
                height: largeImage.height,
                transform: translate,
                WebkitTransform: translate,
                msTransform: translate,
                pointerEvents: 'none'
            };
            var compositeImageStyle = (0, _objectAssign2.default)({}, imageStyle, computedImageStyle);

            var component = _react2.default.createElement(
                'div',
                {
                    className: containerClassName,
                    style: compositeContainerStyle
                },
                _react2.default.createElement('img', {
                    alt: largeImage.alt,
                    className: imageClassName,
                    src: largeImage.src,
                    srcSet: largeImage.srcSet,
                    sizes: largeImage.sizes,
                    style: compositeImageStyle,
                    onLoad: onLoad,
                    onError: onError
                })
            );

            if (isLazyLoaded) {
                return isVisible ? component : null;
            }

            return component;
        }
    }]);

    return _class;
}(_react2.default.Component);

_class.displayName = 'EnlargedImage';
_class.defaultProps = {
    fadeDurationInMs: 0,
    isLazyLoaded: true
};
_class.propTypes = {
    containerClassName: _propTypes2.default.string,
    containerStyle: _propTypes2.default.object,
    cursorOffset: _Point2.default,
    position: _Point2.default,
    fadeDurationInMs: _propTypes2.default.number,
    imageClassName: _propTypes2.default.string,
    imageStyle: _propTypes2.default.object,
    isActive: _propTypes2.default.bool,
    isLazyLoaded: _propTypes2.default.bool,
    largeImage: _ImageShape2.default,
    smallImage: _ImageShape2.default,
    imagePosition: _propTypes2.default.oneOf([_constants.ENLARGED_IMAGE_POSITION.beside, _constants.ENLARGED_IMAGE_POSITION.over])
};
exports.default = _class;
